### ISSUE TITLE
[react/no-typos] unexpected error "`'react'` imported without a local `React` binding"

### ISSUE THREAD

**Issue by SuperOleg39** [yannickcr/eslint-plugin-react#2857]
Hello!

I have a simple `vendor.ts` file:

```ts
import 'react';
import 'react-dom';
```
With `eslint-plugin-react`, i have this error - `'react' imported without a local 'React' binding  react/no-typos`

`react/no-typos` rule responsibility - prevent typos, but `vendor.ts` doesn't have any.

Any chances to turn off this error, or fix [this branch](https://github.com/yannickcr/eslint-plugin-react/commit/5bf1f0f625820f9daf5ab0c45602b959229478cb#diff-250c0de4be0e45f06179dda7f56a1c3c927bd97993c184db29881dc56e1c3da3R181) conditionals?

**Comment by ljharb** [comment]
Unless you have a strange build of react or react-dom, these imports will have no effect, so the warning seems correct.

However, if you are confident you can ignore the warning, then that is done with eslint override comments in-file, or by disabling the rule for the file in “overrides” in your eslint config.

**Comment by SuperOleg39** [comment]
Yes, i create independent bundle from this file.

`import 'react'` - is fully valid code, like any other imports.
Of course i can disable this rule for one line of my code, but i think open-source - it's about make things better)

If i will use `React` in this file, this error makes sense.
But even is this case, is `react/react-in-jsx-scope` rule designed for this case?

**Comment by ljharb** [comment]
No, absolutely not - having `import 'react'` execute side effects is not what react *does*, so nothing in the ecosystem is likely to be designed for a case where you've somehow made it have side effects.

**Comment by SuperOleg39** [comment]
What about this example:

```ts
import 'style.css'
```

This code have side effects?

**Comment by ljharb** [comment]
Yes, I would assume it does, since i'm unfamiliar with the specifier and you have no bindings - but react *specifically* does not have any, so `import 'react'` is useless.

**Comment by SuperOleg39** [comment]
About `import​ ​'style.css'` example - i want to say when this code can have side effects - if bundler have specific configuration, for creating separate css chunk.
In other cases, it can be a css modules, and user need use styles manually - `import styles from 'style.css'`, and nobody mark first case as error.

And bundler configuration for `import 'whatever-library'` and creating vendor chunk from this imports has no conceptual difference between css import.

Of course, with webpack, splitChunks is preferred for vendor chunk creating, but my project have some legacy features.

So, if css import for bundler side effect - is valid case, why react import - invalid?

**Comment by ljharb** [comment]
Because `import 'react'` isn't *doing* anything. It doesn't create a React global, the react package has no side effects, literally the only impact it's having is that it's increasing your bundle size with a bunch of code *that you can't ever use*.

If, as you're implying, you've configured your bundler to do something abnormal for `import 'react'` - like create a global - or if you're using it for managing bundle splitting, then you've chosen to do something atypical/nonstandard that deviates from how the ecosystem works, so it's very expected you'll have to work around ecosystem tools.

**Comment by SuperOleg39** [comment]
> increasing your bundle size with a bunch of code that you can't ever use

This is not entirely true, i already use this code, all `react` imports inside other bundles, resolving from vendor bundle.
`autodll-webpack-plugin` (with `DllPlugin` under the hood) make this possible, and I cannot say that this is something non-standard.

Webpack `optimization.splitChunks` option, for example with [similar configuration](https://github.com/vercel/next.js/blob/8221c180a55828b39a7e1e01811365bfaaa0dd19/packages/next/build/webpack-config.ts#L457), will make same thing, increasing bundle size with a react code, just without explicit import.

**Comment by ljharb** [comment]
Indeed it is non-standard to use an explicit side-effecting import to do bundle-splitting; usually bundler config decides this, and everything that needs React, say, would `import React from 'react'`.

This seems answered to me; you're welcome to override the rule in this case if it doesn't fit your individually distinct use case.

**Comment by receter** [comment]
The `react/no-typos` rule is (incorrectly?) triggered when using a valid side-effect-only import in a TypeScript declaration (.d.ts) file. This pattern is necessary for module augmentation in TypeScript, but the rule expects a local React binding even when one is not required or intended.
